pkg/cacheutil/memcached_client.go

```diff
@@ -0,0 +1,382 @@
+	}
+}
+
+func (c *MemcachedClientConfig) validate() error {
+	if len(c.Addrs) == 0 {
+		return errMemcachedConfigNoAddrs
+	}
+
+	return nil
+}
+
+type memcachedClient struct {
+	logger   log.Logger
+	config   MemcachedClientConfig
+	client   memcachedClientBackend
+	selector *MemcachedJumpHashSelector
+
+	// DNS provider used to keep the memcached servers list updated.
+	provider *dns.Provider
+
+	// Channel used to notify internal goroutines when they should quit.
+	stop chan struct{}
+
+	// Channel used to enqueue async operations.
+	asyncQueue   chan func()
+	asyncWorkers sync.WaitGroup
+
+	// Channel used to enqueue get multi operations.
+	getMultiQueue   chan *memcachedGetMultiBatch
+	getMultiWorkers sync.WaitGroup
+}
+
+type memcachedGetMultiBatch struct {
+	keys    []string
+	results chan<- *memcachedGetMultiResult
+}
+
+type memcachedGetMultiResult struct {
+	items map[string]*memcache.Item
+	err   error
+}
+
+// NewMemcachedClient makes a new MemcachedClient.
+func NewMemcachedClient(logger log.Logger, provider *dns.Provider, config MemcachedClientConfig) (MemcachedClient, error) {
+	// We use a custom servers selector in order to use a jump hash
+	// for servers selection.
+	selector := &MemcachedJumpHashSelector{}
+
+	client := memcache.NewFromSelector(selector)
+	client.Timeout = config.Timeout
+	client.MaxIdleConns = config.MaxIdleConnections
+
+	return newMemcachedClient(logger, client, selector, provider, config)
+}
+
+func newMemcachedClient(logger log.Logger, client memcachedClientBackend, selector *MemcachedJumpHashSelector, provider *dns.Provider, config MemcachedClientConfig) (MemcachedClient, error) {
+	config.applyDefaults()
+	if err := config.validate(); err != nil {
+		return nil, err
+	}
+
+	c := &memcachedClient{
+		logger:        logger,
+		config:        config,
+		client:        client,
+		selector:      selector,
+		provider:      provider,
+		asyncQueue:    make(chan func(), config.MaxAsyncBufferSize),
+		getMultiQueue: make(chan *memcachedGetMultiBatch),
+		stop:          make(chan struct{}, 1),
+	}
+
+	// As soon as the client is created it must ensure that memcached server
+	// addresses are resolved, so we're going to trigger an initial addresses
+	// resolution here.
+	if err := c.resolveAddrs(); err != nil {
+		return nil, err
+	}
+
+	go c.resolveAddrsLoop()
+
+	// Start a number of goroutines - processing async operations - equal
+	// to the max concurrency we have.
+	c.asyncWorkers.Add(c.config.MaxAsyncConcurrency)
+	for i := 0; i < c.config.MaxAsyncConcurrency; i++ {
+		go c.asyncQueueProcessLoop()
+	}
+
+	// Start a number of goroutines - processing get multi batch operations - equal
+	// to the max concurrency we have.
+	c.getMultiWorkers.Add(c.config.MaxGetMultiBatchConcurrency)
+	for i := 0; i < c.config.MaxGetMultiBatchConcurrency; i++ {
+		go c.getMultiQueueProcessLoop()
+	}
+
+	return c, nil
+}
+
+func (c *memcachedClient) Stop() {
+	close(c.stop)
+
+	// Wait until all workers have terminated.
```

Comment:
do we also want to wait for the resolve addrs loop?

Reply:
I thought about it, but I don't see any benefit waiting for it. Do you have any thought on it?